File: backend/src/aerisun/api/admin/site_config.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from aerisun.core.db import get_session
from aerisun.domain.iam.models import AdminUser
from aerisun.domain.ops.config_revisions import capture_config_resource, create_config_revision
from aerisun.domain.site_config.models import NavItem, PageCopy, PageDisplayOption, Poem, SocialLink
from aerisun.domain.site_config.service import (
    attach_site_profile_id,
    get_community_config_admin,
    get_site_profile_admin,
    reorder_nav_items_admin,
    site_profile_scoped_query,
    update_community_config_admin,
    update_site_profile_admin,
)

from .content import build_crud_router
from .deps import get_current_admin
from .schemas import (
    CommunityConfigAdminRead,
    CommunityConfigUpdate,
    NavItemAdminRead,
    NavItemCreate,
    NavItemUpdate,
    NavReorderItem,
    PageCopyAdminRead,
    PageCopyCreate,
    PageCopyUpdate,
    PageDisplayOptionAdminRead,
    PageDisplayOptionCreate,
    PageDisplayOptionUpdate,
    PoemAdminRead,
    PoemCreate,
    PoemUpdate,
    SiteProfileAdminRead,
    SiteProfileUpdate,
    SocialLinkAdminRead,
    SocialLinkCreate,
    SocialLinkUpdate,
)

router = APIRouter(prefix="/site-config", tags=["admin-site-config"])


@router.get("/profile", response_model=SiteProfileAdminRead, summary="获取站点资料")
def get_profile(
    _admin: AdminUser = Depends(get_current_admin),
    session: Session = Depends(get_session),
) -> Any:
    return get_site_profile_admin(session)


@router.put("/profile", response_model=SiteProfileAdminRead, summary="更新站点资料")
def update_profile(
    payload: SiteProfileUpdate,
    _admin: AdminUser = Depends(get_current_admin),
    session: Session = Depends(get_session),
) -> Any:
    before_snapshot = capture_config_resource(session, "site.profile")
    result = update_site_profile_admin(session, payload)
    after_snapshot = capture_config_resource(session, "site.profile")
    create_config_revision(
        session,
        actor_id=_admin.id,
        resource_key="site.profile",
        operation="update",
        before_snapshot=before_snapshot,
        after_snapshot=after_snapshot,
    )
    return result


@router.get("/community-config", response_model=CommunityConfigAdminRead, summary="获取社区评论配置")
def get_community_config(
    _admin: AdminUser = Depends(get_current_admin),
    session: Session = Depends(get_session),
) -> Any:
    return get_community_config_admin(session)


@router.put("/community-config", response_model=CommunityConfigAdminRead, summary="更新社区评论配置")
def update_community_config(
    payload: CommunityConfigUpdate,
    _admin: AdminUser = Depends(get_current_admin),
    session: Session = Depends(get_session),
) -> Any:
    before_snapshot = capture_config_resource(session, "site.community")
    result = update_community_config_admin(session, payload)
    after_snapshot = capture_config_resource(session, "site.community")
    create_config_revision(
        session,
        actor_id=_admin.id,
        resource_key="site.community",
        operation="update",
        before_snapshot=before_snapshot,
        after_snapshot=after_snapshot,
    )
    return result


social_links_router = build_crud_router(
    SocialLink,
    create_schema=SocialLinkCreate,
    update_schema=SocialLinkUpdate,
    read_schema=SocialLinkAdminRead,
    prefix="/social-links",
    tag="admin-site-config",
    base_query_factory=lambda session: site_profile_scoped_query(session, SocialLink),
    prepare_create_data=attach_site_profile_id,
    config_resource_key="site.social_links",
    capture_before=lambda session: capture_config_resource(session, "site.social_links"),
)

poems_router = build_crud_router(
    Poem,
    create_schema=PoemCreate,
    update_schema=PoemUpdate,
    read_schema=PoemAdminRead,
    prefix="/poems",
    tag="admin-site-config",
    base_query_factory=lambda session: site_profile_scoped_query(session, Poem),
    prepare_create_data=attach_site_profile_id,
    config_resource_key="site.poems",
    capture_before=lambda session: capture_config_resource(session, "site.poems"),
)

page_copy_router = build_crud_router(
    PageCopy,
    create_schema=PageCopyCreate,
    update_schema=PageCopyUpdate,
    read_schema=PageCopyAdminRead,
    prefix="/page-copy",
    tag="admin-site-config",
    config_resource_key="site.pages",
    capture_before=lambda session: capture_config_resource(session, "site.pages"),
)

display_options_router = build_crud_router(
    PageDisplayOption,
    create_schema=PageDisplayOptionCreate,
    update_schema=PageDisplayOptionUpdate,
    read_schema=PageDisplayOptionAdminRead,
    prefix="/display-options",
    tag="admin-site-config",
    config_resource_key="site.pages",
    capture_before=lambda session: capture_config_resource(session, "site.pages"),
)

router.include_router(social_links_router)
router.include_router(poems_router)
router.include_router(page_copy_router)
router.include_router(display_options_router)


@router.put("/nav-items/reorder", response_model=list[NavItemAdminRead], summary="重排导航项顺序")
def reorder_nav_items(
    items: list[NavReorderItem],
    _admin: AdminUser = Depends(get_current_admin),
    session: Session = Depends(get_session),
) -> Any:
    before_snapshot = capture_config_resource(session, "site.navigation")
    result = reorder_nav_items_admin(session, items)
    after_snapshot = capture_config_resource(session, "site.navigation")
    create_config_revision(
        session,
        actor_id=_admin.id,
        resource_key="site.navigation",
        operation="update",
        before_snapshot=before_snapshot,
        after_snapshot=after_snapshot,
    )
    return result


nav_items_router = build_crud_router(
    NavItem,
    create_schema=NavItemCreate,
    update_schema=NavItemUpdate,
    read_schema=NavItemAdminRead,
    prefix="/nav-items",
    tag="admin-site-config",
    base_query_factory=lambda session: site_profile_scoped_query(session, NavItem),
    prepare_create_data=attach_site_profile_id,
    config_resource_key="site.navigation",
    capture_before=lambda session: capture_config_resource(session, "site.navigation"),
)

router.include_router(nav_items_router)

File: admin/src/api/endpoints/site-config.ts
import client from "../client";
import type {
  SiteProfile,
  SiteProfileUpdate,
  SocialLink,
  SocialLinkCreate,
  SocialLinkUpdate,
  Poem,
  PoemCreate,
  PoemUpdate,
  PageCopy,
  PageCopyCreate,
  PageCopyUpdate,
  PageDisplayOption,
  PageDisplayOptionCreate,
  PageDisplayOptionUpdate,
  NavItem,
  NavItemCreate,
  NavItemUpdate,
  CommunityConfig,
  CommunityConfigUpdate,
  PaginatedResponse,
} from "@/types/models";
import { COMMUNITY_CONFIG_ENDPOINTS } from "@/lib/community-config";

// --- Profile ---
export async function getProfile(): Promise<SiteProfile> {
  const res = await client.get("/site-config/profile");
  return res.data;
}

export async function updateProfile(data: SiteProfileUpdate): Promise<SiteProfile> {
  const res = await client.put("/site-config/profile", data);
  return res.data;
}

// --- Social Links ---
export async function listSocialLinks(params?: { page?: number }): Promise<PaginatedResponse<SocialLink>> {
  const res = await client.get("/site-config/social-links/", { params });
  return res.data;
}

export async function createSocialLink(data: SocialLinkCreate): Promise<SocialLink> {
  const res = await client.post("/site-config/social-links/", data);
  return res.data;
}

export async function updateSocialLink(id: string, data: SocialLinkUpdate): Promise<SocialLink> {
  const res = await client.put(`/site-config/social-links/${id}`, data);
  return res.data;
}

export async function deleteSocialLink(id: string): Promise<void> {
  await client.delete(`/site-config/social-links/${id}`);
}

// --- Poems ---
export async function listPoems(params?: { page?: number }): Promise<PaginatedResponse<Poem>> {
  const res = await client.get("/site-config/poems/", { params });
  return res.data;
}

export async function createPoem(data: PoemCreate): Promise<Poem> {
  const res = await client.post("/site-config/poems/", data);
  return res.data;
}

export async function updatePoem(id: string, data: PoemUpdate): Promise<Poem> {
  const res = await client.put(`/site-config/poems/${id}`, data);
  return res.data;
}

export async function deletePoem(id: string): Promise<void> {
  await client.delete(`/site-config/poems/${id}`);
}

// --- PageCopy ---
export async function listPageCopy(params?: { page?: number }): Promise<PaginatedResponse<PageCopy>> {
  const res = await client.get("/site-config/page-copy/", { params });
  return res.data;
}

export async function createPageCopy(data: PageCopyCreate): Promise<PageCopy> {
  const res = await client.post("/site-config/page-copy/", data);
  return res.data;
}

export async function updatePageCopy(id: string, data: PageCopyUpdate): Promise<PageCopy> {
  const res = await client.put(`/site-config/page-copy/${id}`, data);
  return res.data;
}

export async function deletePageCopy(id: string): Promise<void> {
  await client.delete(`/site-config/page-copy/${id}`);
}

// --- Display Options ---
export async function listDisplayOptions(params?: { page?: number }): Promise<PaginatedResponse<PageDisplayOption>> {
  const res = await client.get("/site-config/display-options/", { params });
  return res.data;
}

export async function createDisplayOption(data: PageDisplayOptionCreate): Promise<PageDisplayOption> {
  const res = await client.post("/site-config/display-options/", data);
  return res.data;
}

export async function updateDisplayOption(id: string, data: PageDisplayOptionUpdate): Promise<PageDisplayOption> {
  const res = await client.put(`/site-config/display-options/${id}`, data);
  return res.data;
}

export async function deleteDisplayOption(id: string): Promise<void> {
  await client.delete(`/site-config/display-options/${id}`);
}

// --- Nav Items ---
export async function listNavItems(params?: { page?: number }): Promise<PaginatedResponse<NavItem>> {
  const res = await client.get("/site-config/nav-items/", { params });
  return res.data;
}

export async function createNavItem(data: NavItemCreate): Promise<NavItem> {
  const res = await client.post("/site-config/nav-items/", data);
  return res.data;
}

export async function updateNavItem(id: string, data: NavItemUpdate): Promise<NavItem> {
  const res = await client.put(`/site-config/nav-items/${id}`, data);
  return res.data;
}

export async function deleteNavItem(id: string): Promise<void> {
  await client.delete(`/site-config/nav-items/${id}`);
}

export async function reorderNavItems(items: { id: string; order_index: number }[]): Promise<void> {
  await client.put("/site-config/nav-items/reorder", items);
}

// --- Community / Comment System ---
async function requestCommunityConfig<T>(method: "get" | "put", data?: CommunityConfigUpdate): Promise<T> {
  let lastError: unknown = null;

  for (const endpoint of COMMUNITY_CONFIG_ENDPOINTS) {
    try {
      const res = method === "get"
        ? await client.get(endpoint)
        : await client.put(endpoint, data);
      return (res.data?.item ?? res.data) as T;
    } catch (error) {
      lastError = error;
    }
  }

  throw lastError instanceof Error ? lastError : new Error("Unable to load community config");
}

export async function getCommunityConfig(): Promise<CommunityConfig> {
  return requestCommunityConfig<CommunityConfig>("get");
}

export async function updateCommunityConfig(data: CommunityConfigUpdate): Promise<CommunityConfig> {
  return requestCommunityConfig<CommunityConfig>("put", data);
}
